Add ASE calculator for xtp using pybind11

NicoRenaud opened the pull request:
This PR adds an ASE calculator for xtp that allows to easily set up calculation from python that we have set up with @baumeier. The call to xtp is done via pybind11 interface and not system call. Creating an ASE calculator allows to use the features offered by ASE such as geometry optimization, molecular dynamics, etc ... The code works with/without MKL support but was only tested on Ubuntu so far. 

I would love to start a discussion to see if we could merge that into Votca.

junghans commented on .github/workflows/continuous-integration-workflow.yml:
```diff
@@ -155,6 +155,9 @@ jobs:
       - name: Install
         run: sudo LD_LIBRARY_PATH=$LD_LIBRARY_PATH cmake --build builddir --target install
+      - name: Test pyxtp
+        run: apt-get install -y python3-xml2dict ase
```
That is in the container now!

NicoRenaud replied:
removed

junghans commented on tools/src/libtools/CMakeLists.txt:
```diff
@@ -28,11 +28,11 @@ set_target_properties(votca_tools PROPERTIES SOVERSION ${SOVERSION})
-  endif()
+#  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+#    target_link_options(votca_tools PUBLIC "-Wl,--no-as-needed")
+#  endif()
```
Remo

junghans commented on xtp/CMakeLists.txt:
```diff
@@ -43,6 +43,11 @@ if(USE_CUDA)
   set_package_properties(CUDA PROPERTIES TYPE REQUIRED PURPOSE "Used for GPU acceleration")
 endif(USE_CUDA)
 
+find_package(Python 3 COMPONENTS Interpreter Development)
```
I think we are looking for the python unconditionally in the root cmake now.

NicoRenaud replied:
removed

junghans commented on xtp/src/CMakeLists.txt:
```diff
@@ -1,3 +1,8 @@
 add_subdirectory(tools)
-add_subdirectory(tests)
+if(ENABLE_TESTING)
+  add_subdirectory(tests)
```
I think we are testing for enable_testing in the directory now.

NicoRenaud replied:
Changed that and indeed the tests were not build since ENABLE_TESTING was not defined. The test_orca.cc had some issues with the QMPacakgeFactory call due to the initial merge. I fix that as well

junghans commented on xtp/src/pyxtp/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+find_package(Python 3 REQUIRED COMPONENTS Interpreter Development)
```
We are already checked for that above

junghans commented on xtp/src/pyxtp/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+find_package(Python 3 REQUIRED COMPONENTS Interpreter Development)
+find_package(pybind11 REQUIRED)
```
Check a second time here….

junghans commented on xtp/src/pyxtp/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+file(GLOB_RECURSE BIND_SOURCES binds/*.cc)
+
+pybind11_add_module(xtp_binds ${BIND_SOURCES})
+target_include_directories(xtp_binds PUBLIC binds/include)
```
Hmm why public? And it should be Build_interface only unless we install the headers

NicoRenaud replied:
I honestly don't know. Felipe was apparently the engineer who touched this line and he's not working here for a while now. Would you recommend:

```
target_include_directories(xtp_binds PRIVATE $<BUILD_INTERFACE:binds/include>)
```

junghans replied:
Yes, please

NicoRenaud replied:
done

baumeier commented on xtp/include/votca/xtp/toolfactory.h:
```diff
@@ -32,13 +32,22 @@ namespace xtp {
+//  public:
+//   QMToolFactory() { this->RegisterAll(); };
+
+ // master branch
```
remove comments

NicoRenaud replied:
thanks !

baumeier commented on xtp/include/votca/xtp/qmpackagefactory.h:
```diff
@@ -31,7 +31,14 @@ namespace votca {
+
+//  public:
+//   QMPackageFactory() { this->RegisterAll(); };
+
```
remove comments

baumeier commented on xtp/src/pyxtp/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+install(FILES ${PYXTP_FILES} DESTINATION ${PYTHON_INSTDIR}/pyxtp)
+
+
+# add_test(NAME integration_test_xtp_binds COMMAND ${Python_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/examples.py WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
```
remove commented code?

NicoRenaud replied:
Done thanks

NicoRenaud commented:
I've fixed all your comments so far thanks ! I'm facing an issue though. In `xtp_tools.cc` [here](https://github.com/NicoRenaud/votca/blob/8bc442e68ec650a9e8c16d04ba4aaebc700ca7a8/xtp/src/pyxtp/binds/xtp_tools.cc#L40) it seems that `dftgwbse` is not registered and I get the error message `factory key dftgwbse not found` I don't understand what's happening here I must admit. Any ideas ?

baumeier commented:
@NicoRenaud Does it only affect the `dftgwbse` calculator or all?

NicoRenaud commented:
@baumeier I think all. I’ve printed the keys that I get from the .getKeys() method and there are none. Even after calling RegisterAll()

baumeier commented:
Very strange. Could it be some missing includes from qmtools.h or toolfactory?

NicoRenaud commented:
@baumeier  I don't think it's a missing include but I'll check and keep you posted

baumeier commented:
@NicoRenaud I have no idea why that was introduced and by whom! @JensWehner?

NicoRenaud commented:
QMTools and also QMToolFactory are only used for the pybind interface. So maybe I can simply refactor QMToolFactory so that it works the same as CalculatorFactory (which seems to be ok for the pybind interface) ?

NicoRenaud commented:
@baumeier I've removed the friend class `QMTools`, that is apparently difficult to access from pybind. I've refactored `QMToolFactory` that is declared in `xtp/src/libxtp/factories/toolfactory.cc` and `xtp/src/libxtp/factories/toolfactory.h` so that we don't need the friend class to call `RegisterAll()` in `XtpTools` defined in `xtp/src/tools/xtp_tools.cc`. The code compiles and the tests are passing. Could you take a look at the files mentionned above see if there is anything that is not ok ?

NicoRenaud commented:
@junghans it seems that the installation of `ASE` and `xml2dict` doesn't work when using intel (see [here](https://github.com/NicoRenaud/votca/actions/runs/5054480819/jobs/9069464753)). Similarly `pyxtp` doesn't seem to be installed properly when using clang (see [there](https://github.com/NicoRenaud/votca/actions/runs/5055085935/jobs/9072280189)). Do you have any idea what could be happening there ?

junghans commented:
> @junghans it seems that the installation of `ASE` and `xml2dict` doesn't work when using intel (see [here](https://github.com/NicoRenaud/votca/actions/runs/5054480819/jobs/9069464753)). Could it be because we installed the dependencies on the native python and use intel python instead ?
> 
> It also seems that XTP is not built in some cases (see e.g. [there](https://github.com/NicoRenaud/votca/actions/runs/5055085935/jobs/9072280189)). In those cases pytest fails of course.

Hmm, I think both issues could be solved by running pytest through CMake! For the latter the test should not be run of xtp isn't build. For the former I think you are mixing python interpreters somehow.

NicoRenaud commented:
> Hmm, I think both issues could be solved by running pytest through CMake! 

That would be great ! I'll look into it

> For the latter the test should not be run of xtp isn't build. 

I'm trying to only test pyxtp only if the `minimal==False` in the setup. I guess that should work

>For the former I think you are mixing python interpreters somehow.

I think so too ... I'll try to fix that

NicoRenaud commented:
I've added the test to CMAKE but I can't set the PYTHONPATH correctly and the test fails because it can't find the pyxtp library. The corresponding CMAKE code is [here ](https://github.com/NicoRenaud/votca/blob/0ed2719594c99e307dd0712325999ba43f909715/xtp/src/pyxtp/CMakeLists.txt#L21) @junghans  and @baumeier any idea on how to properly set the path ?

junghans commented:
> I've added the test to CMAKE but I can't set the PYTHONPATH correctly and the test fails because it can't find the pyxtp library. The corresponding CMAKE code is [here ](https://github.com/NicoRenaud/votca/blob/0ed2719594c99e307dd0712325999ba43f909715/xtp/src/pyxtp/CMakeLists.txt#L21) @junghans and @baumeier any idea on how to properly set the path ?

You can do:
```
set_property(TEST ${TESTNAME} PROPERTY ENVIRONMENT PYTHONPATH=${CMAKE_CURRENT_BINARY_DIR}/foo/bar/)
```

NicoRenaud commented:
Thanks @junghans the  main issue is that the CI was running ctest before installing the library .... I'm still having some issues with python interpreter using intel though it seems. I'll fix that next

NicoRenaud commented:
@junghans I've tried setting the path to the CURRENT_BIN_DIR or CURRENT_SOURCE_DIR but I have issues with the python libraries and the bindings bot being found. But is it not better to run test with the installed libraries, and therefore define the PYTHONPATH to PYTHON_INSTDIR ? 

As for the issue I have with intel could it be because we install the intel tool chains after installing all the dependencies in build_env/fedora ?

junghans commented:
> @junghans I've tried setting the path to the CURRENT_BIN_DIR or CURRENT_SOURCE_DIR but I have issues with the python libraries and the bindings bot being found. But is it not better to run test with the installed libraries, and therefore define the PYTHONPATH to PYTHON_INSTDIR ?

 No, the idea is to test things before installing  a possibly broken library. You can run `ctest -V -R <name of the test>`  by hand and adjust the `PYTHON_PATH` in the there. You might have to list 2 directories in there.
> 
> As for the issue I have with intel could it be because we install the intel tool chains after installing all the dependencies in build_env/fedora ?

I have no idea... But the intel compiler comes with it's own python and hence I guess there is a mix up somewhere.

junghans commented:
Now one can build xtp without pybind11 anymore is that intended?

NicoRenaud commented:
If someone would like to have xtp but only the C++ and not the python version that could be possible. But that might add unnecessary complexity ... But pybind is always installed  with all the dependencies through the `buildend` files no ?

NicoRenaud commented:
One thing we could do is to not use `intelpython` it's usually done when sourcing the setvars.sh from oneapi. However I can't find where that is done in the CI. It seems that opeapi is installed via dnf but that's it. Am i missing something ?

junghans commented:
> If someone would like to have xtp but only the C++ and not the python version that could be possible. But that might add unnecessary complexity ... But pybind is always installed with all the dependencies through the `buildend` files no ?

This is more a question for @baumeier do we want to force all users to have pybind to use xtp.

junghans commented:
> One thing we could do is to not use `intelpython` it's usually done when sourcing the setvars.sh from oneapi. However I can't find where that is done in the CI. It seems that opeapi is installed via dnf but that's it. Am i missing something ?

Yes, we only set the PATH and LD_LIBRARY_PATH.

baumeier commented:
> > If someone would like to have xtp but only the C++ and not the python version that could be possible. But that might add unnecessary complexity ... But pybind is always installed with all the dependencies through the `buildend` files no ?
> 
> This is more a question for @baumeier do we want to force all users to have pybind to use xtp.

That’s a good question. We maybe need to consider something mid-/longterm. I had the idea of moving some of the C++ code to python, from things like making a histogram to maybe even the job control for `xtp_parallel`, eventually. Then I suppose we will have to force pybind, or one will have some limited functionality only. 

I can’t tell how difficult it is from a build environment/cmake perspective to make the python bindings optional for now. Any idea?

junghans commented on xtp/CMakeLists.txt:
```diff
@@ -43,6 +43,10 @@ if(USE_CUDA)
   set_package_properties(CUDA PROPERTIES TYPE REQUIRED PURPOSE "Used for GPU acceleration")
 endif(USE_CUDA)
 
+set_package_properties(Python PROPERTIES TYPE RECOMMENDED PURPOSE "Used to build xtp python bindings")
```
Remove and update doc for Python usage in main `CMakeLists.txt`.

junghans commented on xtp/src/pyxtp/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+  execute_process(
+    COMMAND
+      ${Python_EXECUTABLE} -c
+      "import distutils.sysconfig as cg; print(cg.get_python_lib(plat_specific=True, standard_lib=False))"
```
Triggers a warning in python3.10, see #924.

junghans replied:
@NicoRenaud we need to fix this, from [PEP 632](https://peps.python.org/pep-0632/)
```
Code that imports distutils will no longer work from Python 3.12.
```

junghans replied:
The code we need to replace is the following:
```python
import distutils.sysconfig as cg
print(cg.get_python_lib(1,0,prefix='${CMAKE_INSTALL_PREFIX}'))
```

junghans replied:
https://discourse.cmake.org/t/findpython3-how-to-specify-local-installation-directory-for-python-module/3580

junghans replied:
I guess this would be semi-reasonable:
```cmake
if (NOT CMAKE_INSTALL_PYTHON_LIBDIR)
  set(CMAKE_INSTALL_PYTHON_LIBDIR
    "${CMAKE_INSTALL_LIBDIR}/python${major}.${minor}/site-packages")
endif ()
```

junghans commented on xtp/src/pyxtp/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+install(FILES ${PYXTP_FILES} DESTINATION ${PYTHON_INSTDIR}/pyxtp)
+
+foreach(_PYFILE ${PYXTP_FILES})
+  get_filename_component(_FILE "${_PYFILE}" NAME)
```
We could make that part easier if the `py` script would be in `pyxtp` instead of `pyxtp/pyxtp`.

junghans replied:
@NicoRenaud Any opinion if we should move this?

junghans replied:
It might make things to cluttered in this dir, so let's leave it.

junghans commented on xtp/src/pyxtp/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+add_dependencies(xtp_binds xtp_binary_scripts)
+
+if(BUILD_TESTING)
+  add_test(NAME integration_test_xtp_binds COMMAND ${Python_EXECUTABLE} -m pytest ${CMAKE_CURRENT_BIANRY_DIR}/.. )#/test/ WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/test/)
```
Remove comment!

junghans commented on xtp/src/CMakeLists.txt:
```diff
@@ -1,3 +1,8 @@
 add_subdirectory(tests)
+add_subdirectory(pyxtp)
+
+# if(TARGET Python::Module AND pybind11_FOUND)
```
Something like this would allow to make `pyxtp` optional

junghans commented:
I am not 100% sure what this error means:
```
../../../../../../../usr/lib/python3.11/site-packages/_pytest/cacheprovider.py:432
543
  /usr/lib/python3.11/site-packages/_pytest/cacheprovider.py:432: PytestCacheWarning: could not create cache path /.pytest_cache/v/cache/nodeids
```
but I get it locally too.

junghans commented on CMakeLists.txt:
```diff
@@ -138,7 +138,7 @@ else()
 # iie methods needs python 3.5+
-find_package(Python 3.5 REQUIRED COMPONENTS Interpreter)
+find_package(Python 3.5 REQUIRED COMPONENTS Interpreter Development)
 set_package_properties(Python PROPERTIES TYPE OPTIONAL PURPOSE "Used in csg scripts and docu")
```
Just add a comment about pyxtp here.

NicoRenaud commented:
Thanks @junghans we can now run the test without having to install the library first !



 We are still facing two major problems:

1. the test with intel are failing because `ase` and `xml2dict` are not found. I don't think that we ever even install intelpython nor do we add it to the PATH. So I'm not sure what's happening here and I can't figure it out
2. The sphinx build fails because the call to `xtp_parallel` in `QMMM_LAMMPS.ipynb` fails. This has to do once more with the `xtp` key not being registered with `QMPackageFactory` and it's friend class `QMPackages.` I've started refactoring that in a separate branch because that's a bit more work compared to the same issue we had with `QMToolFactory` and `QMTools`. I also find it strange that xtp_parallel is now failing since I didn't touch it .... 

I'll be in holiday most of next week and I will have less time in June to continue this project. 

I hope we can manage to finish this at some point

Thanks for your help

NicoRenaud commented:
> > > If someone would like to have xtp but only the C++ and not the python version that could be possible. But that might add unnecessary complexity ... But pybind is always installed with all the dependencies through the `buildend` files no ?
> > 
> > 
> > This is more a question for @baumeier do we want to force all users to have pybind to use xtp.
> 
> That’s a good question. We maybe need to consider something mid-/longterm. I had the idea of moving some of the C++ code to python, from things like making a histogram to maybe even the job control for `xtp_parallel`, eventually. Then I suppose we will have to force pybind, or one will have some limited functionality only.
> 
> I can’t tell how difficult it is from a build environment/cmake perspective to make the python bindings optional for now. Any idea?

I personally wouldn't make the building of the python binding optional, especially if there are plans to keep extending the python capabilities. That will introduce even more complexity in the build and I don't see what we would gain from it.

junghans commented:
We should also have CMake check for `xmltodict`.

junghans commented:
> Thanks @junghans we can now run the test without having to install the library first !
> 
> We are still facing two major problems:
> 
> 1. the test with intel are failing because `ase` and `xml2dict` are not found. I don't think that we ever even install intelpython nor do we add it to the PATH. So I'm not sure what's happening here and I can't figure it out
> 2. The sphinx build fails because the call to `xtp_parallel` in `QMMM_LAMMPS.ipynb` fails. This has to do once more with the `xtp` key not being registered with `QMPackageFactory` and it's friend class `QMPackages.` I've started refactoring that in a separate branch because that's a bit more work compared to the same issue we had with `QMToolFactory` and `QMTools`. I also find it strange that xtp_parallel is now failing since I didn't touch it ....
> 
> I'll be in holiday most of next week and I will have less time in June to continue this project.
> 
> I hope we can manage to finish this at some point
> 
> Thanks for your help

I will look at the intel issue while you are gone.

NicoRenaud commented:
I think all the tests are passing now see [here](https://github.com/NicoRenaud/votca/actions/runs/5129688156) ! Thanks @junghans for fixing the dependencies install on intel. 

I do still have two caveats:

1. I am not testing the calculations of forces through pyxtp because the test was timing out on a few install in DEBUG mode. We are using a standard ASE routine to compute the forces so it might be ok to assume that they work ...
2. I don't have permission to deploy the documentation through sphinx. So I've disabled that for now in the CI. I don't know if it's because I'm not part of the votca organization or because of something else. @junghans could you help me with that ?

junghans commented:
> I think all the tests are passing now see [here](https://github.com/NicoRenaud/votca/actions/runs/5129688156) ! Thanks @junghans for fixing the dependencies install on intel.
> 
> I do still have two caveats:
> 
> 1. I am not testing the calculations of forces through pyxtp because the test was timing out on a few install in DEBUG mode. We are using a standard ASE routine to compute the forces so it might be ok to assume that it works ...

You could exclude the test in CMake for debug builds.

> 2. I don't have permission to deploy the documentation through sphinx. So I've disabled that for now in the CI. I don't know if it's because I'm not part of the votca organization or because of something else. @junghans could you help me with that ?

Let me have a look!

junghans commented:
@NicoRenaud can you touch your branch so things get triggered as you?

junghans commented:
@NicoRenaud please also have a look at Codacy warnings: https://app.codacy.com/gh/votca/votca/pullRequest?prid=11759848

NicoRenaud commented:
@junghans I've fixed most of the codacy issues. There are still some left that have to do with the use of xml files that I'm not sure how to handle (or even if we do want to fix them). I still can't deploy the sphinx doc though  ...

NicoRenaud commented:
I think the deploy issue comes from my side. I'll fix it later today

junghans commented:
@NicoRenaud what seems to be the problem? All CI job passed! For external PR that don't come from a organization member the trigger is always manually for security reasons, but that is just GitHub policy.

NicoRenaud commented:
@junghans  i think the problem is that I ve been logging in with a token and I think it doesn’t have the right scope.

NicoRenaud commented:
> > @junghans i think the problem is that I ve been logging in with a token and I think it doesn’t have the right scope.
> 
> Not sure what you mean, the deploy happens automatically once the PR is merged!

Ok I don’t have to fix it then ! :) 

For the code itself the main change is that we had to remove the singleton construction from all the factories. @baumeier (and @JensWehner) could you check if that is ok ?

junghans commented:
@votca-bot changelog: add python bindings for xtp

junghans commented:
> > > @junghans i think the problem is that I ve been logging in with a token and I think it doesn’t have the right scope.
> > 
> > 
> > Not sure what you mean, the deploy happens automatically once the PR is merged!
> 
> Ok I don’t have to fix it then ! :)

Ok, I understand the deploy issue now, the CI fails in your fork, but works here. The reason is that you are working on `master` (your `master`), which we hadn't anticipated in the CI. I will see if I can fix that.

junghans commented:
@votca-bot copyright
